### ISSUE TITLE
Follow the W3C JSON forms spec

### ISSUE THREAD

**Issue by LinusU** [expressjs/multer#129]
Currently we are are using `qs` to do the final transform on our fields. The drawback with this is that they are all buffered in a different format than the final, so when `req` is exposed to something that wants to check previous fields, it can complicated things.

I also believe that we are relying on undocumented features of `qs` which might be a bad idea.

I think that we should use a [W3C JSON forms spec](http://www.w3.org/TR/html-json-forms/) compatible, incrementally capable parser. This should be very easy to build ourselves if we don't find a good one.

Related issue hapijs/qs#63

**Comment by LinusU** [comment]
I've made a [repository containing some tests](https://github.com/LinusU/testdata-w3c-json-form).

**Comment by jpfluger** [comment]
Your repo follows the [w3c examples](http://www.w3.org/TR/html-json-forms/). Nice.

**Comment by jpfluger** [comment]
So for example 6 in the w3c, submitting a form where `enctype=application/json` qs outputs:

```json
{"name":"Bender","hind":"Bitable","shiny":"on","wow":{"such":{"deep":[{"much":{"power":{"[!]":"Amaze"}}}]}}}
```
Obviously, it's missing the nulls.... You thinking to adapt [qs/parse.js](https://github.com/hapijs/qs/blob/master/lib/parse.js), starting with `parseObject`?

**Comment by LinusU** [comment]
It's not really clear unless you read the spec in full but the `null`s are actually only there because of `JSON.stringify` and it really should be a sparse array.

The other problem is that `qs` isn't incremental and it would be a big plus for us to have that. (being able to expose the fields parsed in the callbacks for file handlers)

I've fixed the problems in #131 now :)

**Comment by LinusU** [comment]
This is now fixed in multer `1.0.0`
